server/service/orbit_client.go

```diff
@@ -124,16 +136,87 @@ func NewOrbitClient(
 	}
 
 	nodeKeyFilePath := filepath.Join(rootDir, constant.OrbitNodeKeyFileName)
+
+	ctx, cancelFunc := context.WithCancel(context.Background())
+
 	return &OrbitClient{
-		nodeKeyFilePath:   nodeKeyFilePath,
-		baseClient:        bc,
-		enrollSecret:      enrollSecret,
-		hostInfo:          orbitHostInfo,
-		enrolled:          false,
-		onGetConfigErrFns: onGetConfigErrFns,
+		nodeKeyFilePath:          nodeKeyFilePath,
+		baseClient:               bc,
+		enrollSecret:             enrollSecret,
+		hostInfo:                 orbitHostInfo,
+		enrolled:                 false,
+		onGetConfigErrFns:        onGetConfigErrFns,
+		ReceiverUpdateInterval:   defaultOrbitConfigReceiverInterval,
+		ReceiverUpdateContext:    ctx,
+		ReceiverUpdateCancelFunc: cancelFunc,
 	}, nil
 }
 
+func (oc *OrbitClient) RunConfigReceivers() error {
+	config, err := oc.GetConfig()
+	if err != nil {
+		return fmt.Errorf("RunReceivers get config: %w", err)
+	}
+
+	var errs []error
+	var errMu sync.Mutex
+	var wg sync.WaitGroup
+	wg.Add(len(oc.ConfigReceivers))
+
+	for _, receiver := range oc.ConfigReceivers {
+		receiver := receiver
+		go func() {
```

Comment:
Main concern here is if any config receiver blocks for some reason, the whole `RunConfigReceivers` call blocks. We should probably pass a cancelable context specific to running the receivers, and have it timeout after some duration (and created from the parent context so it is also canceled if the parent context is canceled), but it is also tricky to determine a reasonable timeout duration.

Reply:
I had that thought as well, but didn't want to start refactoring all of the receivers until I got some eyes on the changes. I think sending a cancelable context with a timeout to all receivers makes a ton of sense 👍

Reply:
That being said, I don't think there's anything stopping a single fetcher from holding up the rest in the current design either

Reply:
Indeed!